File: Algorithms/merge-sort/merge_sort.py
#merge sort in python
# Written by Moses K.
# l   m   r
# 1 2 3 4 6 

def merge(array, left, mid, right):
    sizeA = mid - left + 1
    sizeB = right - mid

    A = [sizeA]
    B = [sizeB]

    for i in range(sizeA):
        A[i] = array[left + i]
    
    for j in range(sizeB):
        B[j] = array[mid + 1 + j]
    
    i = 0
    j = 0
    k = left

    while i < sizeA and j < sizeB:
        if A[i] <= B[j]:
            array[k] = A[i]
            i += 1
        else:
            array[k] = B[j]
            j += 1
        k += 1
    
    while i < sizeA:
        array[k] = A[i]
        i += 1
        k += 1

    while j < sizeB:
        array[k] = B[j]
        j += 1
        k += 1


def mergeSort(array, left, right):
    if left < right:
        mid = left + right // 2
        mergeSort(array, left, mid)
        mergeSort(array, mid + 1, right)
        merge(array, left, mid, right)


# test the function
res = [23,2,45,0,12,5,6]
size = len(res)
print(mergeSort(res, 0, size - 1))

File: Algorithms/merge-sort/merge_sort.c
#include <stdio.h>

// merge sort algorithm
/*
            How it workds
    - The array is recursively divided into 2 halves
    - Each half is then sorted recursively
    - the sorted halves are then merged back together

    Time & Space complexity:
        time: O(n log n)
        space: 
*/

//merge two subarrays 
void merge(int arr[], int left, int mid, int right){
    int n1 = mid - left + 1;
    int n2 = right - mid;

    //create temp arrays
    int L[n1], R[n2];

    //copy data to temp arrays
    for(int i = 0; i < n1; i++){
        L[i] = arr[left + i];
    }

    for(int j = 0; j < n2; j++){
        R[j] = arr[mid + 1 + j];
    }

    //merge them back int arr[left...righ]
    int i = 0;
    int j = 0;
    int k = left;
    while(i < n1 && j < n2){
        if(L[i] <= R[j]){
            arr[k] = L[i];
            i++;
        } else{
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    //copy the raiming elements of L[] if any
    while(i < n1){
        arr[k] = L[i];
        i++;
        k++;
    }

    //copy the raiming elements of R[] if any
    while(j < n2){
        arr[k] = R[j];
        j++;
        k++;
    }
}

void mergeSort(int arr[], int left, int right){
    if(left < right){
        //find the mid point
        int mid = left + (right - left) / 2;

        //sort the first and second half
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);

        //merge the sorted halves
        merge(arr, left, mid, right);

    }
}
void printArray(int arr[], int size){
    for(int i = 0; i < size; i++){
        printf("%d ", arr[i]);
    }
    printf("\n");
}

int main(){
    int arr[] = {12,11,13,5,6,7};
    int arr_size = sizeof(arr) / sizeof(arr[0]);

    printf("Given array is \n");
    printArray(arr, arr_size);

    mergeSort(arr, 0, arr_size - 1);

    printf("\nSorted array is \n");
    printArray(arr, arr_size);
    return 0;
}